### PR TITLE
feat: v2 api endpoints and hooks to list private and public event types of user

### PR DESCRIPTION
**Pull request by supalarry**
## What does this PR do?

Private event types are event types that user sees in his / her dashboard.
Public ones are event types that other users see when viewing user public booking page e.g. i.cal.com/user

1. Adds 2 new endpoints to the [event-types.controller.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-82354a0146bf90a12f9079e54f26ea1384bf698aec49361ed83f09de18960d0e) to fetch private and public event types of a user.
2. For private event types - extract a new lib function [getEventTypesPublic.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-a13a731e1fc2f99862b81c7ccfc91f079b6656bdcd8574de1865871bc755a341) that is re-used in the new endpoint and also in the web app [users-public-view.getServerSideProps.tsx](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-3b142cee714eff52af99f20840c2df286db6b8eedd9eaf15274fd5b3c8411314).
3. For public event types - extract a new lib function [getEventTypesByViewer.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-c72adbb7f441f901ab56a7a1d56b7ef14c1bd8d4af732e0652ac201301ac5819) that is used in the new endpoint and also in web app [getByViewer.handler.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-5294731b960e76c72f504ce7e1381d51e5c054131976fe7f43bb38f485d76aa6).
4. Adds new hooks [useEventTypesPrivate.ts](https://github.com/calcom/cal.com/pull/14192/files#diff-641628268c5ee492057d7bf1d55613cf03946894d80c34e25b2ef7dd0ccce4e8) and [useEventTypesPublic.ts](https://github.com/calcom/cal.com/pull/14192/files#diff-647b3057e2eb3c11b12431d0a97d088dbec8a517058b4079aaab449f9d133803).
5. Organizes lib event-types helpers by moving them to "@calcom/lib/event-types"
6. Tests for fetching private and public event types: [event-types.controller.e2e-spec.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-6afea2aadb85fab021adbe8be6aae0067bd4c4a8cf0dab1cf59a45f804d45562)

See my comment below https://github.com/calcom/cal.com/pull/14192#issuecomment-2016869761 about test fix.

### Private event types
Request URL: http://localhost:5555/api/v2/event-types
Function invoked: packages/lib/event-types/getEventTypesByViewer.ts
Response:
```
{
  "status": "success",
  "data": {
    "eventTypeGroups": [
      {
        "teamId": null,
        "bookerUrl": "http://localhost:3000",
        "membershipRole": null,
        "profile": {
          "slug": "fbdycqblyi",
          "name": "John Jones",
          "image": "http://localhost:3000/fbdycqblyi/avatar.png"
        },
        "eventTypes": [
          {
            "id": 1168,
            "teamId": null,
            "schedulingType": null,
            "userId": 47,
            "metadata": null,
            "description": null,
            "hidden": false,
            "slug": "thirty-minutes",
            "length": 30,
            "title": "30 Minutes",
            "requiresConfirmation": false,
            "position": 0,
            "offsetStart": 0,
            "profileId": null,
            "eventName": null,
            "parentId": null,
            "timeZone": null,
            "periodType": "UNLIMITED",
            "periodStartDate": null,
            "periodEndDate": null,
            "periodDays": null,
            "periodCountCalendarDays": null,
            "lockTimeZoneToggleOnBookingPage": false,
            "requiresBookerEmailVerification": false,
            "disableGuests": false,
            "hideCalendarNotes": false,
            "minimumBookingNotice": 120,
            "beforeEventBuffer": 0,
            "afterEventBuffer": 0,
            "seatsPerTimeSlot": null,
            "onlyShowFirstAvailableSlot": false,
            "seatsShowAttendees": false,
            "seatsShowAvailabilityCount": true,
            "scheduleId": null,
            "price": 0,
            "currency": "usd",
            "slotInterval": null,
            "successRedirectUrl": null,
            "isInstantEvent": false,
            "assignAllTeamMembers": false,
            "recurringEvent": null,
            "hashedLink": null,
            "children": [],
            "hosts": [],
            "userIds": [
              47
            ]
          },
          {
            "id": 1169,
            "teamId": null,
            "schedulingType": null,
            "userId": 47,
            "metadata": null,
            "description": null,
            "hidden": false,
            "slug": "sixty-minutes-video",
            "length": 60,
            "title": "60 Minutes",
            "requiresConfirmation": false,
            "position": 0,
            "offsetStart": 0,
            "profileId": null,
            "eventName": null,
            "parentId": null,
            "timeZone": null,
            "periodType": "UNLIMITED",
            "periodStartDate": null,
            "periodEndDate": null,
            "periodDays": null,
            "periodCountCalendarDays": null,
            "lockTimeZoneToggleOnBookingPage": false,
            "requiresBookerEmailVerification": false,
            "disableGuests": false,
            "hideCalendarNotes": false,
            "minimumBookingNotice": 120,
            "beforeEventBuffer": 0,
            "afterEventBuffer": 0,
            "seatsPerTimeSlot": null,
            "onlyShowFirstAvailableSlot": false,
            "seatsShowAttendees": false,
            "seatsShowAvailabilityCount": true,
            "scheduleId": null,
            "price": 0,
            "currency": "usd",
            "slotInterval": null,
            "successRedirectUrl": null,
            "isInstantEvent": false,
            "assignAllTeamMembers": false,
            "recurringEvent": null,
            "hashedLink": null,
            "children": [],
            "hosts": [],
            "userIds": [
              47
            ]
          },
          {
            "id": 1170,
            "teamId": null,
            "schedulingType": null,
            "userId": 47,
            "metadata": null,
            "description": null,
            "hidden": false,
            "slug": "sixty-minutes",
            "length": 60,
            "title": "60 Minutes",
            "requiresConfirmation": false,
            "position": 0,
            "offsetStart": 0,
            "profileId": null,
            "eventName": null,
            "parentId": null,
            "timeZone": null,
            "periodType": "UNLIMITED",
            "periodStartDate": null,
            "periodEndDate": null,
            "periodDays": null,
            "periodCountCalendarDays": null,
            "lockTimeZoneToggleOnBookingPage": false,
            "requiresBookerEmailVerification": false,
            "disableGuests": false,
            "hideCalendarNotes": false,
            "minimumBookingNotice": 120,
            "beforeEventBuffer": 0,
            "afterEventBuffer": 0,
            "seatsPerTimeSlot": null,
            "onlyShowFirstAvailableSlot": false,
            "seatsShowAttendees": false,
            "seatsShowAvailabilityCount": true,
            "scheduleId": null,
            "price": 0,
            "currency": "usd",
            "slotInterval": null,
            "successRedirectUrl": null,
            "isInstantEvent": false,
            "assignAllTeamMembers": false,
            "recurringEvent": null,
            "hashedLink": null,
            "children": [],
            "hosts": [],
            "userIds": [
              47
            ]
          },
          {
            "id": 1171,
            "teamId": null,
            "schedulingType": null,
            "userId": 47,
            "metadata": null,
            "description": null,
            "hidden": false,
            "slug": "thirty-minutes-video",
            "length": 30,
            "title": "30 Minutes",
            "requiresConfirmation": false,
            "position": 0,
            "offsetStart": 0,
            "profileId": null,
            "eventName": null,
            "parentId": null,
            "timeZone": null,
            "periodType": "UNLIMITED",
            "periodStartDate": null,
            "periodEndDate": null,
            "periodDays": null,
            "periodCountCalendarDays": null,
            "lockTimeZoneToggleOnBookingPage": false,
            "requiresBookerEmailVerification": false,
            "disableGuests": false,
            "hideCalendarNotes": false,
            "minimumBookingNotice": 120,
            "beforeEventBuffer": 0,
            "afterEventBuffer": 0,
            "seatsPerTimeSlot": null,
            "onlyShowFirstAvailableSlot": false,
            "seatsShowAttendees": false,
            "seatsShowAvailabilityCount": true,
            "scheduleId": null,
            "price": 0,
            "currency": "usd",
            "slotInterval": null,
            "successRedirectUrl": null,
            "isInstantEvent": false,
            "assignAllTeamMembers": false,
            "recurringEvent": null,
            "hashedLink": null,
            "children": [],
            "hosts": [],
            "userIds": [
              47
            ]
          }
        ],
        "metadata": {
          "membershipCount": 1,
          "readOnly": false
        }
      }
    ],
    "profiles": [
      {
        "slug": "fbdycqblyi",
        "name": "John Jones",
        "image": "http://localhost:3000/fbdycqblyi/avatar.png",
        "membershipCount": 1,
        "readOnly": false,
        "teamId": null,
        "membershipRole": null
      }
    ],
    "allUsersAcrossAllEventTypes": {}
  }
}
```

### Public event types
Request URL: http://localhost:5555/api/v2/event-types/{{USERNAME}}/public
Function invoked: packages/lib/event-types/getEventTypesPublic.ts
Response:
```
{
  "status": "success",
  "data": [
    {
      "id": 1176,
      "title": "qc",
      "description": "",
      "length": 15,
      "schedulingType": null,
      "recurringEvent": null,
      "slug": "qc",
      "hidden": false,
      "price": 0,
      "currency": "usd",
      "lockTimeZoneToggleOnBookingPage": false,
      "requiresConfirmation": false,
      "requiresBookerEmailVerification": false,
      "metadata": {},
      "descriptionAsSafeHTML": ""
    },
    {
      "id": 1177,
      "title": "lc",
      "description": "",
      "length": 60,
      "schedulingType": null,
      "recurringEvent": null,
      "slug": "lc",
      "hidden": false,
      "price": 0,
      "currency": "usd",
      "lockTimeZoneToggleOnBookingPage": false,
      "requiresConfirmation": false,
      "requiresBookerEmailVerification": false,
      "metadata": {},
      "descriptionAsSafeHTML": ""
    }
  ]
}
```

### CODE REVIEW

**Comment by supalarry** [comment]
## Here is what I don't get:

"yarn test" was failing in"ui" and  "app-store"  with "crypto" package not being resolved:

Screenshot 2024-03-24 at 17 39 16


Screenshot 2024-03-24 at 17 39 12

## Solution
I updated "test-setup.ts" of "ui" and "app-store" in commit [fix: yarn test](https://github.com/calcom/cal.com/pull/14192/commits/347f3cdeef333cf608ce3126ec8118bcd69bae21) by mocking "@calcom/prisma" and "@calcom/lib/server/repository/profile" and it works - see CI and I ran "yarn test" locally and it passes.

Screenshot 2024-03-24 at 17 52 52

## Why though?
"@calcom/lib/server/repository/profile" appears in the newly extracted [getEventTypesByViewer.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-c72adbb7f441f901ab56a7a1d56b7ef14c1bd8d4af732e0652ac201301ac5819) and the newly extracted [getEventTypesPublic.ts](https://github.com/calcom/cal.com/compare/lauris/pla-14-api-endpoint-hook-to-list-event-types-of-user?expand=1#diff-a13a731e1fc2f99862b81c7ccfc91f079b6656bdcd8574de1865871bc755a341) has prisma.

If we don't add mocks in the "##Solution" section but just stop exporting the new function in "lib/index" then tests also pass with:
Screenshot 2024-03-24 at 17 55 34

Tests are now green with the new mocks, but it feels "dirty" to me because i don't know why it works:
Screenshot 2024-03-24 at 17 57 10

File: packages/platform/atoms/hooks/event-types/useEventTypesPublic.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import http from "../../lib/http";
+
+export const QUERY_KEY = "get-private-events";
+export const useEventTypesPublic = (username: string) => {
+  const endpoint = new URL(BASE_URL);
+
+  endpoint.pathname = `api/${API_VERSION}/${V2_ENDPOINTS.eventTypes}/${username}/public`;
+
+  return useQuery({
+    queryKey: [QUERY_KEY],
```
Comment (ThyMinimalDev) [inline packages/platform/atoms/hooks/event-types/useEventTypesPublic.ts:16]: `queryKey: [QUERY_KEY, username]`

Comment (ThyMinimalDev) [reply packages/platform/atoms/hooks/event-types/useEventTypesPublic.ts:16]: react-query needs to be aware of the username to know that when the props change it should fetch

Comment (supalarry) [reply packages/platform/atoms/hooks/event-types/useEventTypesPublic.ts]: Resolved in [refactor: track dynamic prop in react query](https://github.com/calcom/cal.com/pull/14192/commits/a3d534587f0c354af73a6149356836807f853b98)

**Review by ThyMinimalDev — APPROVED** [review]

**Pull request merged**